Postponed warning message fails with class cast exception

### Expected Behavior

```
Caused by: java.lang.ClassCastException: class io.micronaut.annotation.processing.visitor.JavaClassElement cannot be cast to class javax.lang.model.element.Element (io.micronaut.annotation.processing.visitor.JavaClassElement is in unnamed module of loader java.net.URLClassLoader @3ac99c68; javax.lang.model.element.Element is in module java.compiler of loader 'platform')
        at io.micronaut.annotation.processing.BeanDefinitionInjectProcessor.process(BeanDefinitionInjectProcessor.java:234)
        at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
        at org.gradle.api.internal.tasks.compile.processing.DynamicProcessor.process(DynamicProcessor.java:52)
        at org.gradle.api.internal.tasks.compile.processing.DelegatingProcessor.process(DelegatingProcessor.java:62)
        at org.gradle.api.internal.tasks.compile.processing.TimeTrackingProcessor.access$401(TimeTrackingProcessor.java:37)
```

### Actual Behaviour

_No response_

### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

_No response_

### Version

4.8.x

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
@@ -15,6 +15,7 @@
  */
 package io.micronaut.annotation.processing;
 
+import io.micronaut.annotation.processing.visitor.AbstractJavaElement;
 import io.micronaut.annotation.processing.visitor.JavaClassElement;
 import io.micronaut.annotation.processing.visitor.JavaNativeElement;
 import io.micronaut.context.annotation.ClassImport;
@@ -41,6 +42,7 @@ import io.micronaut.inject.writer.BeanDefinitionWriter;
 import javax.annotation.processing.ProcessingEnvironment;
 import javax.annotation.processing.RoundEnvironment;
 import javax.annotation.processing.SupportedOptions;
+import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.Name;
@@ -244,8 +246,11 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
         */
         if (processingOver) {
             for (Map.Entry<String, PostponeToNextRoundException> e : postponed.entrySet()) {
-                javaVisitorContext.warn("Bean definition generation [" + e.getKey() + "] skipped from processing because of prior error: [" + e.getValue().getPath() + "]." +
-                    " This error is normally due to missing classes on the classpath. Verify the compilation classpath is correct to resolve the problem.", (Element) e.getValue().getErrorElement());
+                String className = e.getKey();
+                Object errorElement = e.getValue().getErrorElement();
+                Element failedElement = resolvedFailedElement(errorElement);
+                javaVisitorContext.warn("Bean definition generation [" + className + "] skipped from processing because of prior error: [" + e.getValue().getPath() + "]." +
+                    " This error is normally due to missing classes on the classpath. Verify the compilation classpath is correct to resolve the problem.", failedElement);
             }
 
             try {
@@ -275,6 +280,20 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
         }
 
         return false;
+    }
+
+    private static Element resolvedFailedElement(Object errorElement) {
+        Element failedElement;
+        if (errorElement instanceof Element el) {
+            failedElement = el;
+        } else if (errorElement instanceof JavaNativeElement jne) {
+            failedElement = jne.element();
+        } else if (errorElement instanceof AbstractJavaElement aje) {
+            failedElement = aje.getNativeType().element();
+        } else {
+            failedElement = null;
+        }
+        return failedElement;
     }
 
     /**

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/BeanDefinitionInjectProcessor.java
@@ -15,7 +15,6 @@
  */
 package io.micronaut.annotation.processing;
 
-import io.micronaut.annotation.processing.visitor.AbstractJavaElement;
 import io.micronaut.annotation.processing.visitor.JavaClassElement;
 import io.micronaut.annotation.processing.visitor.JavaNativeElement;
 import io.micronaut.context.annotation.ClassImport;
@@ -42,7 +41,6 @@ import io.micronaut.inject.writer.BeanDefinitionWriter;
 import javax.annotation.processing.ProcessingEnvironment;
 import javax.annotation.processing.RoundEnvironment;
 import javax.annotation.processing.SupportedOptions;
-import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.Name;
@@ -194,9 +192,9 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
                         final TypeElement typeElement = elementUtils.getTypeElement(className);
                         PostponeToNextRoundException nextRoundException = postponed.get(className);
                         if (nextRoundException != null) {
-                            Object errorElement = nextRoundException.getErrorElement();
-                            if (errorElement instanceof Element element) {
-                                AbstractAnnotationMetadataBuilder.CachedAnnotationMetadata cachedAnnotationMetadata = javaVisitorContext.getAnnotationMetadataBuilder().lookupOrBuildForType(element);
+                            Element errorElement = nextRoundException.getNativeErrorElement();
+                            if (errorElement != null) {
+                                AbstractAnnotationMetadataBuilder.CachedAnnotationMetadata cachedAnnotationMetadata = javaVisitorContext.getAnnotationMetadataBuilder().lookupOrBuildForType(errorElement);
                                 if (!cachedAnnotationMetadata.wasCleared()) {
                                     AbstractAnnotationMetadataBuilder.clearMutated(errorElement);
                                 }
@@ -230,7 +228,8 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
                                 }
                             }
                         } catch (ProcessingException ex) {
-                            error(((JavaNativeElement) ex.getOriginatingElement()).element(), ex.getMessage());
+                            JavaNativeElement javaNativeElement = (JavaNativeElement) ex.getOriginatingElement();
+                            error(javaNativeElement != null ? javaNativeElement.element() : null, ex.getMessage());
                         } catch (PostponeToNextRoundException e) {
                             processed.remove(className);
                             postponed.put(className, e);
@@ -247,8 +246,7 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
         if (processingOver) {
             for (Map.Entry<String, PostponeToNextRoundException> e : postponed.entrySet()) {
                 String className = e.getKey();
-                Object errorElement = e.getValue().getErrorElement();
-                Element failedElement = resolvedFailedElement(errorElement);
+                Element failedElement = e.getValue().getNativeErrorElement();
                 javaVisitorContext.warn("Bean definition generation [" + className + "] skipped from processing because of prior error: [" + e.getValue().getPath() + "]." +
                     " This error is normally due to missing classes on the classpath. Verify the compilation classpath is correct to resolve the problem.", failedElement);
             }
@@ -280,20 +278,6 @@ public class BeanDefinitionInjectProcessor extends AbstractInjectAnnotationProce
         }
 
         return false;
-    }
-
-    private static Element resolvedFailedElement(Object errorElement) {
-        Element failedElement;
-        if (errorElement instanceof Element el) {
-            failedElement = el;
-        } else if (errorElement instanceof JavaNativeElement jne) {
-            failedElement = jne.element();
-        } else if (errorElement instanceof AbstractJavaElement aje) {
-            failedElement = aje.getNativeType().element();
-        } else {
-            failedElement = null;
-        }
-        return failedElement;
     }
 
     /**

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/MixinVisitorProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/MixinVisitorProcessor.java
@@ -105,11 +105,11 @@ public class MixinVisitorProcessor extends AbstractInjectAnnotationProcessor {
                     }
                     error(originatingElement.element(), e.getMessage());
                 } catch (PostponeToNextRoundException e) {
-                    postponedTypes.put(mixin.getCanonicalName(), e.getErrorElement());
+                    postponedTypes.put(mixin.getCanonicalName(), e.getNativeErrorElement());
                 } catch (ElementPostponedToNextRoundException e) {
                     Object nativeType = e.getOriginatingElement().getNativeType();
-                    if (nativeType instanceof JavaNativeElement jne) {
-                        Element element = jne.element();
+                    Element element = PostponeToNextRoundException.resolvedFailedElement(nativeType);
+                    if (element != null) {
                         postponedTypes.put(mixin.getCanonicalName(), element);
                     } else {
                         // should never happen.

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/PostponeToNextRoundException.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/PostponeToNextRoundException.java
@@ -15,6 +15,12 @@
  */
 package io.micronaut.annotation.processing;
 
+import io.micronaut.annotation.processing.visitor.AbstractJavaElement;
+import io.micronaut.annotation.processing.visitor.JavaNativeElement;
+import io.micronaut.core.annotation.Nullable;
+
+import javax.lang.model.element.Element;
+
 /**
  * Exception to indicate postponing processing to next round.
  */
@@ -34,6 +40,24 @@ public final class PostponeToNextRoundException extends RuntimeException {
 
     public Object getErrorElement() {
         return errorElement;
+    }
+
+    public @Nullable Element getNativeErrorElement() {
+        return resolvedFailedElement(errorElement);
+    }
+
+    public static Element resolvedFailedElement(Object errorElement) {
+        Element failedElement;
+        if (errorElement instanceof Element el) {
+            failedElement = el;
+        } else if (errorElement instanceof JavaNativeElement jne) {
+            failedElement = jne.element();
+        } else if (errorElement instanceof AbstractJavaElement aje) {
+            failedElement = aje.getNativeType().element();
+        } else {
+            failedElement = null;
+        }
+        return failedElement;
     }
 
     public String getPath() {

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/TypeElementVisitorProcessor.java
@@ -291,9 +291,10 @@ public class TypeElementVisitorProcessor extends AbstractInjectAnnotationProcess
                             }
                             error(originatingElement.element(), e.getMessage());
                         } catch (PostponeToNextRoundException e) {
-                            postponeElement(javaClassElement, e.getErrorElement(), e);
+                            postponeElement(javaClassElement, e.getNativeErrorElement(), e);
                         } catch (ElementPostponedToNextRoundException e) {
-                            postponeElement(javaClassElement, e.getOriginatingElement(), e);
+                            Element element = PostponeToNextRoundException.resolvedFailedElement(e.getOriginatingElement());
+                            postponeElement(javaClassElement, element, e);
                         }
                     }
                 }
@@ -330,23 +331,12 @@ public class TypeElementVisitorProcessor extends AbstractInjectAnnotationProcess
         return false;
     }
 
-    private <T extends Throwable> void postponeElement(JavaClassElement javaClassElement, Object originalElement, T e) throws T {
-        Element postponedElement;
-        if (originalElement instanceof Element element) {
-            postponedElement = element;
-        } else if (originalElement instanceof io.micronaut.inject.ast.Element element) {
-            Object nativeType = element.getNativeType();
-            if (nativeType instanceof JavaNativeElement jne) {
-                postponedElement = jne.element();
-            } else {
-                // should never happen.
-                throw e;
-            }
-        } else {
+    private <T extends Throwable> void postponeElement(JavaClassElement javaClassElement, Element originalElement, T e) throws T {
+        if (originalElement == null) {
             // should never happen.
             throw e;
         }
-        postponedTypes.put(javaClassElement.getCanonicalName(), postponedElement);
+        postponedTypes.put(javaClassElement.getCanonicalName(), originalElement);
     }
 
     private void visitClass(LoadedVisitor visitor, JavaClassElement classElement) {

--- a/inject-java/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/beans/BeanDefinitionSpec.groovy
@@ -16,6 +16,25 @@ import test.another.BeanWithPackagePrivate
 
 class BeanDefinitionSpec extends AbstractTypeElementSpec {
 
+    void "test bean with unresolved type"() {
+        when:
+        buildBeanDefinition('test.Unresolved', '''
+package test;
+
+import jakarta.inject.Singleton;
+
+@Singleton
+class Test {
+    Test(Unresolved unresolved) {}
+}
+''')
+
+        then:
+        def e = thrown(RuntimeException)
+        e.message.contains('cannot find symbol')
+
+    }
+
     void "test getTypeString for format #format"() {
         given:
         def definition = buildBeanDefinition('typestring.Test', '''

